lua-mode.el

```diff
@@ -1814,6 +1817,105 @@ If `lua-process' is nil or dead, start a new process first."
+               "local top_ctx = clone(_G)"
+               ;; But then we take all matches of local x = require y from
+               ;; our code buffer and stuff them into the completion table too.
+               ,@(mapcar (lambda (l)
+                           (format "top_ctx['%s'] = require(%s)"
+                                   (car l) (cadr l))) libs)
+
+               ;; recursively delve into context based on input_parts
+               "local function cpl_for(input_parts, ctx, prefixes)"
+               "  if type(ctx) ~= \"table\" then return {} end"
+               "  if #input_parts == 0 and ctx ~= top_ctx then"
+               "    return ctx"
+               "  elseif #input_parts == 1 then" ; the last segment of input
+               "    local matches = {}"
+               "    for k in pairs(ctx) do"
+               "      if k:find('^' .. input_parts[1]) then"
+               "        local parts = clone(prefixes)"
+               "        table.insert(parts, k)"
+               "        table.insert(matches, table.concat(parts, '.'))"
+               "      end"
+               "    end"
+               "    return matches"
+               "  else" ; more segments of input remain; descend into ctx table
+               "    local token1 = table.remove(input_parts, 1)"
+               "    table.insert(prefixes, first_part)"
+               "    return cpl_for(input_parts, ctx[token1], prefixes)"
+               "  end"
+               "end"
+
+               ;; this is a table of the segments of the input
+               "local input = {"
+               ,@(mapcar (apply-partially 'format "'%s',")
+                         (split-string expr "\\.")) "}"
+               ;; spit it out to a file; lua-mode can't send data back to emacs
+               ,(format "local f = io.open('%s', 'w')" out-file)
+               "for _,l in ipairs(cpl_for(input, top_ctx, {})) do"
+               "  f:write(l .. string.char(10))"
+               "end"
+               "f:close()"
+               "end") "\n"))
+
+(defvar lua-local-require-completions nil
+  "During completion, scan file for local require calls for context.
+
+Defaults to nil because this will cause code to be loaded during completion.
+Loading arbitrary code can have unexpected side-effects, so use with caution.")
+
+(defvar lua-local-require-regexp
+  "^local\\s-+\\(\\w+\\)\\s-*=\\s-*require[( ]+\\([^ )\n]+\\)"
+  "A regexp to match lines where a library is required and put in a local.")
+
+(defun lua-local-libs ()
+  "Find all modules loaded with require which are stored in locals.
+
+Returns a list of lists where the car is the local name and the cadr
+is the string that is passed to require."
+  (save-excursion
+    (when lua-local-require-completions
+      (let ((libs nil))
+        (goto-char (point-min))
+        ;; find each match of "local x = require y" and save for later
+        (while (search-forward-regexp lua-local-require-regexp nil t)
```

Comment:
It's better to use `lua-find-regexp` here, as it will skip expressions found in strings and comments.